File: LibVLCSharp.Avalonia.FuncUI/VideoView.cs
using System.Runtime.InteropServices;
using Avalonia;
using Avalonia.Controls;
using Avalonia.Data;
using Avalonia.Platform;
using LibVLCSharp.Shared;

namespace LibVLCSharp.Avalonia.FuncUI;

public enum VideoState
{
    Play,
    Pause,
    Stop
}

/// <summary>
///     Avalonia VideoView for Windows, Linux and Mac.
/// </summary>
public class VideoView : NativeControlHost
{
    /// <summary>
    ///     Media Data Bound property
    /// </summary>
    /// <summary>
    ///     Defines the <see cref="Media" /> property.
    /// </summary>
    public static readonly DirectProperty<VideoView, Media?> MediaProperty =
        AvaloniaProperty.RegisterDirect<VideoView, Media?>(
            nameof(Media),
            o => o.Media,
            (o, v) => o.Media = v,
            defaultBindingMode: BindingMode.TwoWay);

    /// <summary>
    ///     VideoState Data Bound property
    /// </summary>
    /// <summary>
    ///     Defines the <see cref="VideoState" /> property.
    /// </summary>
    public static readonly DirectProperty<VideoView, VideoState> VideoStateProperty =
        AvaloniaProperty.RegisterDirect<VideoView, VideoState>(
            nameof(VideoState),
            o => o.VideoState,
            (o, v) => o.VideoState = v,
            defaultBindingMode: BindingMode.TwoWay);

    /// <summary>
    ///     MediaPlayer Data Bound property
    /// </summary>
    /// <summary>
    ///     Defines the <see cref="MediaPlayer" /> property.
    /// </summary>
    public static readonly DirectProperty<VideoView, MediaPlayer?> MediaPlayerProperty =
        AvaloniaProperty.RegisterDirect<VideoView, MediaPlayer?>(
            nameof(MediaPlayer),
            o => o.MediaPlayer,
            (o, v) => o.MediaPlayer = v,
            defaultBindingMode: BindingMode.TwoWay);

    private MediaPlayer? _mediaPlayer;
    private IPlatformHandle? _platformHandle;


    private VideoState _videoState = VideoState.Stop;

    public Media? Media
    {
        get => MediaPlayer?.Media;
        set
        {
            if (MediaPlayer is null) return;
            MediaPlayer.Media = value;
        }
    }

    public VideoState VideoState
    {
        get => _videoState;
        set
        {
            switch (value)
            {
                case VideoState.Play:
                    Play();
                    break;
                case VideoState.Pause:
                    Pause();
                    break;
                case VideoState.Stop:
                    Stop();
                    break;
                default:
                    throw new ArgumentOutOfRangeException();
            }
        }
    }

    /// <summary>
    ///     Gets or sets the MediaPlayer that will be displayed.
    /// </summary>
    public MediaPlayer? MediaPlayer
    {
        get => _mediaPlayer;
        set
        {
            if (ReferenceEquals(_mediaPlayer, value)) return;

            Detach();
            _mediaPlayer = value;
            Attach();
        }
    }

    private void Play()
    {
        if (MediaPlayer is null || Media is null) return;

        if (MediaPlayer.Hwnd == IntPtr.Zero && MediaPlayer.XWindow == 0U &&
            MediaPlayer.NsObject == IntPtr.Zero) Attach();

        Console.WriteLine("Play");
        MediaPlayer?.Play(Media);
        _videoState = VideoState.Play;
    }

    private void Pause()
    {
        if (Media != null && MediaPlayer is {IsPlaying: true}) MediaPlayer?.Pause();

        Console.WriteLine("Pause");
        _videoState = VideoState.Pause;
    }

    private void Stop()
    {
        if (Media != null && MediaPlayer is {IsPlaying: true}) MediaPlayer?.Stop();

        Console.WriteLine("Stop");
        _videoState = VideoState.Stop;
    }

    private void Attach()
    {
        if (_mediaPlayer == null || _platformHandle == null || !IsInitialized)
            return;

        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
            _mediaPlayer.Hwnd = _platformHandle.Handle;
        else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
            _mediaPlayer.XWindow = (uint) _platformHandle.Handle;
        else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX)) _mediaPlayer.NsObject = _platformHandle.Handle;
    }

    private void Detach()
    {
        if (_mediaPlayer == null)
            return;

        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
            _mediaPlayer.Hwnd = IntPtr.Zero;
        else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
            _mediaPlayer.XWindow = 0;
        else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX)) _mediaPlayer.NsObject = IntPtr.Zero;
    }

    /// <inheritdoc />
    protected override IPlatformHandle CreateNativeControlCore(IPlatformHandle parent)
    {
        _platformHandle = base.CreateNativeControlCore(parent);

        if (_mediaPlayer == null)
            return _platformHandle;

        Attach();

        return _platformHandle;
    }

    /// <inheritdoc />
    protected override void DestroyNativeControlCore(IPlatformHandle control)
    {
        Detach();
        base.DestroyNativeControlCore(control);
        _platformHandle = null;
    }
}

File: Gui/Extensions/LibVLCSharp/Avalonia/FuncUI/VideoView.fs
namespace LibVLCSharp.Avalonia.FuncUI

open System
open System.Runtime.InteropServices
open Avalonia
open Avalonia.Controls
open Avalonia.Data
open Avalonia.FuncUI.Builder
open Avalonia.FuncUI.Types
open Avalonia.Platform
open LibVLCSharp.Shared

type VideoState =
    | Play
    | Pause
    | Stop

type VideoView() =
    inherit NativeControlHost()
    with
        // ---- Private Properties ----

        let mutable _mediaPlayer: MediaPlayer = null

        let mutable _platformHandle: IPlatformHandle =
            null

        let mutable _videoState: VideoState = Stop


        // ---- Member Properties ----

        /// <summary>
        ///     Gets or sets the MediaPlayer that will be displayed.
        /// </summary>
        member this.MediaPlayer
            with get (): MediaPlayer = _mediaPlayer
            and set value = _mediaPlayer <- value

        member this.Media
            with get (): Media = _mediaPlayer.Media
            and set value =
                if not <| isNull _mediaPlayer then
                    _mediaPlayer.Media <- value

        member this.VideoState
            with get (): VideoState = _videoState
            and set value =
                match value with
                | Play -> this.Play()
                | Pause -> this.Pause()
                | Stop -> this.Stop()


        // ---- Avalonia Properties ----
        static member MediaProperty: DirectProperty<VideoView, Media> =
            AvaloniaProperty.RegisterDirect<VideoView, Media>
                ( nameof Media
                ,  fun (o: VideoView) -> o.Media
                ,  fun (o: VideoView) (v: Media) ->
                    o.Media <- v
                ,  defaultBindingMode = BindingMode.TwoWay
                )
                
        static member MediaPlayerProperty: DirectProperty<VideoView, MediaPlayer> =
            AvaloniaProperty.RegisterDirect<VideoView, MediaPlayer>
                ( nameof MediaPlayer
                ,  fun (o: VideoView) -> o.MediaPlayer
                ,  fun (o: VideoView) (v: MediaPlayer) ->
                    o.MediaPlayer <- v
                ,  defaultBindingMode = BindingMode.TwoWay
                )
                
        static member VideoStateProperty: DirectProperty<VideoView, VideoState> =
            AvaloniaProperty.RegisterDirect<VideoView, VideoState>
                ( nameof VideoState
                ,  fun (o: VideoView) -> o.VideoState
                ,  fun (o: VideoView) (v: VideoState) ->
                    o.VideoState <- v
                ,  defaultBindingMode = BindingMode.TwoWay
                )


        // ---- Action Functions ----

        member this.Play() =
            if isNull this.MediaPlayer || isNull this.Media then
                ()

            this.Attach()
            this.MediaPlayer.Play this.Media |> ignore
            _videoState <- VideoState.Play

        member this.Pause() =
            if not <| isNull this.Media
               && not <| isNull this.Media
               && this.MediaPlayer.IsPlaying then
                this.MediaPlayer.Pause()
                _videoState <- VideoState.Pause

        member this.Stop() =
            if not <| isNull this.Media
               && not <| isNull this.Media
               && this.MediaPlayer.IsPlaying then
                this.MediaPlayer.Stop()
                _videoState <- VideoState.Stop


        // ---- Core ----

        member this.Attach() =
            if isNull _mediaPlayer
               || isNull _platformHandle
               || not this.IsInitialized then
                ()
            else if RuntimeInformation.IsOSPlatform OSPlatform.Windows then
                _mediaPlayer.Hwnd <- _platformHandle.Handle
            else if RuntimeInformation.IsOSPlatform OSPlatform.Linux then
                _mediaPlayer.XWindow <- uint _platformHandle.Handle
            else if RuntimeInformation.IsOSPlatform OSPlatform.OSX then
                _mediaPlayer.NsObject <- _platformHandle.Handle
            else
                failwith "Unsupported platform for the LibVLC VideoView component"


        member this.Detach() =
            if isNull _mediaPlayer then
                ()

            else if RuntimeInformation.IsOSPlatform OSPlatform.Windows then
                _mediaPlayer.Hwnd <- IntPtr.Zero
            else if RuntimeInformation.IsOSPlatform OSPlatform.Linux then
                _mediaPlayer.XWindow <- 0u
            else if RuntimeInformation.IsOSPlatform OSPlatform.OSX then
                _mediaPlayer.NsObject <- IntPtr.Zero
            else
                failwith "Unsupported platform for the LibVLC VideoView component"


        /// <inheritdoc />
        override this.CreateNativeControlCore(parent: IPlatformHandle) =
            _platformHandle <- ``base``.CreateNativeControlCore(parent)

            if not <| isNull _mediaPlayer then
                this.Attach()

            _platformHandle


        /// <inheritdoc />
        override this.DestroyNativeControlCore(control: IPlatformHandle) =
            this.Detach()
            ``base``.DestroyNativeControlCore(control)
            _platformHandle <- null
            
            
        // ---- FuncUI Bindings ----
        
        static member create(attrs: IAttr<VideoView> list) : IView<VideoView> = ViewBuilder.Create<VideoView>(attrs)

        static member mediaPlayer<'t when 't :> VideoView>(value: MediaPlayer) : IAttr<'t> =
            AttrBuilder<'t>
                .CreateProperty<MediaPlayer>(VideoView.MediaPlayerProperty, value, ValueNone)

        static member media<'t when 't :> VideoView>(value: Media) : IAttr<'t> =
            AttrBuilder<'t>
                .CreateProperty<Media>(VideoView.MediaProperty, value, ValueNone)

        static member state<'t when 't :> VideoView>(value: VideoState) : IAttr<'t> =
            AttrBuilder<'t>
                .CreateProperty<VideoState>(VideoView.VideoStateProperty, value, ValueNone)

        static member onLoad<'t when 't :> VideoView>(func: MediaPlayer -> unit, ?subPatchOptions) : IAttr<'t> =
            AttrBuilder.CreateSubscription<MediaPlayer>(
                VideoView.MediaPlayerProperty,
                func,
                ?subPatchOptions = subPatchOptions
            )
